server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1680,6 +1682,7 @@ public void close(String reason, boolean flushEngine) throws IOException {
                 final Engine engine = this.currentEngineReference.getAndSet(null);
                 try {
                     if (engine != null && flushEngine) {
+                        engine.saveInfoToClosedShardService();
```

Comment:
I think you can avoid changing Elasticsearch core and avoid introducing an empty saveInfoToClosedShardService(). You can just introduce the saving logic in the `closeNoLock()` function of the stateless engine.

Reply:
I tried out your suggestion, but I realized I can't rely on the `flushEngine==true` here for regular shard closure -- it sounds like `flushEngine` is [only true when closing an index or shutting down the node](https://github.com/elastic/elasticsearch/blob/a5a4b9e529b66279dd9e7967304d7214d2e36c0b/server/src/main/java/org/elasticsearch/index/IndexService.java#L592-L593). Which means, under normal circumstances, we leave the `SearchEngine` alone (don't call `flushAndClose()`), which makes sense because the `SearchEngine` survives shard closure until reads release reference to it.

Anyway, I believe this leaves me needing elasticsearch core changes after all.

Reply:
Oh wait, it looks like [we call engine.close() farther down](https://github.com/elastic/elasticsearch/blob/a5a4b9e529b66279dd9e7967304d7214d2e36c0b/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1688), not just engine.flushAndClose() here. So both can call ultimately closeNoLock()....